distribution/lib/Standard/Base/0.0.0-dev/src/Meta.enso

```diff
@@ -571,21 +571,74 @@ get_constructor_declaring_type constructor = @Builtin_Method "Meta.get_construct
 
 instrumentor_builtin op args = @Builtin_Method "Meta.instrumentor_builtin"
 
-# Builder to create instrumentation for a function
+## PRIVATE
+   ADVANCED
+
+   Builder to create instrumentation for a function
 type Instrumentor
+    ## PRIVATE
     Value impl
+
+    ## PRIVATE
+       ADVANCED
+
+       Registers callback to be executed at the begining of node/expression
+       execution. The callback `fn` gets UUID of the node/expression that is
+       being executed and can return `Nothing` to continue regular execution
+       or anything else to skip the execution and just return given value.
+
+       Arguments:
+       - fn: The callback function accepting UUID.
     on_enter self (fn : Text -> Any | Nothing) =
         instrumentor_builtin "onEnter" [ self.impl, fn ]
         self
+
+    ## PRIVATE
+       ADVANCED
+
+       Registers callback to be executed when a node/expression evaluation
+       is over. The callback `fn` gets UUID and the computed value. Usually
+       the value is _cached_ and returned from `on_enter` callback next time
+       the same expression is evaluated.
+
+       Arguments:
+       - fn: The callback function accepting UUID and computed value
     on_return self (fn : Text -> Any -> Nothing) =
         instrumentor_builtin "onReturn" [ self.impl, fn ]
         self
-    on_function self (fn : Text -> Any | Nothing) =
+
+    ## PRIVATE
+       ADVANCED
+
+       Registers callback to be executed when a node/expression result is a function.
+       The callback `fn` gets UUID as well as a wrapper around the function. The callback
+       can return `Nothing` (in such case the execution continues) or some other value,
+       which is then returned.
+
+       Arguments:
+       - fn: The callback function accepting UUID and function value
+    on_function self (fn : Text -> Any -> Any | Nothing) =
         instrumentor_builtin "onFunction" [ self.impl, fn ]
         self
+
+    ## PRIVATE
+       ADVANCED
+
+       Registers callback to be executed when there is an exception.
+
+       Arguments:
+       - fn: The callback function accepting any object.
```

Comment:
What is this object that is given to the callback? Is this the exception instance?

Reply:
I have no idea! Do you, know @4e6?

Reply:
Assuming that `on_exception` calls the corresponding method of the existing instrument, it will receive the exception object. For example, it can be a `PanicException`, if the value is evaluated to a panic.

Reply:
Note that `PanicException` is also an internal thing to the runtime and normally Enso code does not see it directly. Our `Panic.catch` allows to view it "through the lens" of `CaughtPanic`. Shall we ensure this method also does this kind of wrapping?

Reply:
Looking at 1baefa575a9ab6f1f1b2d50e7e55b7fcdfc29bbb I am starting to get a feeling that `on_exception` can hardly happen. It is only triggered in case of `InteropException` - e.g. when there is an error in the callback, rather than something we should be dealing in regular IDE execution.

Anyway, let's see what the CI says! If everything is _green_, then I remove the `on_exception` and also `IdExecutionService.onExceptionalCallback` altogether.